#159987404 users should be able to create their profiles

joeeasy opened the pull request:
**What does this PR do?**
Implements a feature that allows users to: 
- view other users profile

- edit their profile

- upload a profile picture

- add a bio

**Description of Task to be completed?**
- install `datauri`
- install `Multer`
- install `Cloudinary`
- write tests for the endpoint
- write a validator middleware

**How should this be manually tested?**
Clone the repository.
Checkout the `ft-users-can-update-profile` branch.
Run `NPM START` on your local machine.
Run database migrations using `npm run migrate`.
Start the server using `NPM START`.

**You can test the following endpoints using postman**
PUT `localhost:8000/api/v1/users/update` to update a user's profile
GET `localhost:8000/api/v1/users/:userId` to view a user's profile
HEADERS
- authorization -> paste your token there
- Content-Type -> application/form-data

REQUIRED FIELDS
-    username: `string`
-   email: `string`


**What are the relevant pivotal tracker stories?**
#159987404

iAmao commented on server/controllers/userController.js:
```diff
@@ -66,6 +70,76 @@ const userController = {
+    multerUploads(req, res, () => {
+      const { currentUser } = req;
+      let image = '';
+      const file = !req.file ? '../../upload/dummy-profile.png' : dataUri(req);
```
I'm guessing, you are uploading a default image if the user did not pass an image. Instead, what we should do is have a default image URL in the column, and not make any call to cloudinary if the user did not upload a file.

joeeasy committed:
feat(user): can update their profile

- users can edit their profile

- users can upload profile picture

- users can add bio

- write mock test for image upload

- remove hound flagging for `console.log` statement

- refactor multer to upload files from memory

- install datauri to convert memory buffers to data url

[(Delivers ) #159987404]

jojitoon commented on swagger.json:
```diff
@@ -331,7 +331,134 @@
         }
       }
-    },
+    
```
remove spaces

iAmao commented on test/users.test.spec.js:
```diff
@@ -0,0 +1,185 @@
+        done();
+      });
+  });
+  it("should return successfully when a user's profile is successfully updated", (done) => {
```
Prefer single quotes over double quotes as suggested by the style guide

iAmao commented on test/users.test.spec.js:
```diff
@@ -0,0 +1,185 @@
+        email: 'joe.test@testmail.com',
+        username: 'joeeasy',
+        bio: 'Hello, here is a brief information about me',
+        interest: 'gaming, movies, music, tech',
```
in future we might want to have this as a table

iAmao commented on test/users.test.spec.js:
```diff
@@ -0,0 +1,185 @@
+      .end((err, res) => {
+        expect(res).to.have.status(200);
+        expect(res.body).to.be.an('object');
+        expect(res.body.data.message).to.equal('Your profile has been updated successfully');
```
You are not asserting that any value actually changed here

iAmao commented on test/users.test.spec.js:
```diff
@@ -0,0 +1,185 @@
+        done();
+      });
+  });
+  it("should return successfully when a user's profile is successfully updated", (done) => {
```
Prefer single quotes over double quotes as suggested by the style guide

iAmao replied:
Also, isn't this a duplicate test?

iAmao commented on test/users.test.spec.js:
```diff
@@ -0,0 +1,185 @@
+        email: 'joe.test@testmail.com',
+        username: 'joeeasy',
+        bio: 'Hello, here is a brief information about me',
+        // interest: 'gaming, movies, music, tech',
```
avoid pushing commented code

iAmao commented on test/users.test.spec.js:
```diff
@@ -0,0 +1,185 @@
+      .type('form')
+      .end((err, res) => {
+        expect(res).to.have.status(200);
+        expect(res.body.data.message).to.equal('Your profile has been updated successfully');
```
There is actually no response validating that this works

iAmao commented on test/users.test.spec.js:
```diff
@@ -0,0 +1,185 @@
+      .set('Authorization', verifiedToken)
+      .end((error, res) => {
+        expect(res).to.have.status(200);
+        assert.isObject(res.body, 'is an object containing the user details');
```
expect the user's details as well

iAmao commented on test/users.test.spec.js:
```diff
@@ -0,0 +1,185 @@
+      .get('/api/v1/users/abcd')
+      .set('Authorization', verifiedToken)
+      .end((message, res) => {
+        expect(res).to.have.status(400);
```
Server should return 404 In this case

iAmao commented on test/users.test.spec.js:
```diff
@@ -0,0 +1,185 @@
+        done();
+      });
+  });
+  it('should not parseInt if Id has alphabet', (done) => {
```
This test is testing for the same case as above. We don't need it

iAmao commented on server/controllers/userController.js:
```diff
@@ -300,6 +303,86 @@ const userController = {
+   * @method updateProfile
+   * @description Allows users to edit their profile
+   * @param {Object} req The request object
+   * @param {Object} res The the response object
```
take out one of the `the`

iAmao commented on server/controllers/userController.js:
```diff
@@ -300,6 +303,86 @@ const userController = {
+      const {
+        firstname, lastname, username, email, bio,
+      } = req.body;
+      const interest = (req.body.interest) ? req.body.interest.split(',') : null;
```
The braces are not needed here though

iAmao commented on server/controllers/userController.js:
```diff
@@ -300,6 +303,86 @@ const userController = {
+  getUserProfile: (req, res) => {
+    const userId = parsedId(req.params.userId);
+    if (!(Number.isInteger(userId))) {
+      return res.status(400).jsend.error({
```
Lets just return 404 for this case

iAmao reviewed (changes requested):
Nice 👍  but, there were some style violations you need to address. Also, some suggestions.
